lib_nn/src/cpp/filt2d/geom/WindowLocation.cpp

```diff
@@ -0,0 +1,99 @@
+
+#include "geom/WindowLocation.hpp"
+
+#include <cassert>
+
+using namespace nn;
+
+ImageVect WindowLocation::InputStart() const {
+  return filter.window.WindowOffset(this->output_coords);
+}
+
+ImageVect WindowLocation::InputEnd() const {
+  return InputStart().add(
+      (filter.window.shape.height - 1) * filter.window.dilation.row,
+      (filter.window.shape.width - 1) * filter.window.dilation.col,
+      filter.window.shape.depth - 1);
+}
+
+ImageVect WindowLocation::InputCoords(const int filter_row,
+                                      const int filter_col,
+                                      const int filter_chan) const {
+  assert(filter_row >= 0);
+  assert(filter_row < filter.window.shape.height);
+  assert(filter_col >= 0);
+  assert(filter_col < filter.window.shape.width);
+  assert(filter_chan >= 0);
+  assert(filter_chan < filter.window.shape.depth);
+
+  const auto in_start = this->InputStart();
+
+  return in_start.add(filter_row * filter.window.dilation.row,
+                      filter_col * filter.window.dilation.col, filter_chan);
+}
+
+int WindowLocation::InputIndex(const int filter_row, const int filter_col,
+                               const int filter_chan) const {
+  assert(filter_row >= 0);
+  assert(filter_row < filter.window.shape.height);
+  assert(filter_col >= 0);
+  assert(filter_col < filter.window.shape.width);
+  assert(filter_chan >= 0);
+  assert(filter_chan < filter.window.shape.depth);
```

Comment:
Is there a reason we mix the terms "chan" and "depth"? It would be nice if it was consistent,

Reply:
this is a good point

Reply:
im going to leave this for now